File: src/main/java/com/vaadin/demo/component/contextmenu/ContextMenuHierarchical.java
package com.vaadin.demo.component.contextmenu;

import com.vaadin.demo.DemoExporter; // hidden-source-line
import com.vaadin.flow.component.grid.Grid;
import com.vaadin.flow.component.grid.contextmenu.GridContextMenu;
import com.vaadin.flow.component.grid.contextmenu.GridMenuItem;
import com.vaadin.flow.component.grid.contextmenu.GridSubMenu;
import com.vaadin.flow.component.html.Div;
import com.vaadin.flow.component.html.Hr;
import com.vaadin.flow.router.Route;

import static org.apache.commons.io.FileUtils.byteCountToDisplaySize;

@Route("context-menu-hierarchical")
public class ContextMenuHierarchical extends Div {

  public ContextMenuHierarchical() {
    Grid<File> grid = new Grid();
    grid.setHeightByRows(true);
    grid.setItems(getFiles());

    grid.addColumn(File::getName)
        .setHeader("Name");
    grid.addColumn(File::getDisplaySize)
        .setHeader("Size");

    // tag::snippet[]
    GridContextMenu<File> menu = grid.addContextMenu();
    // end::snippet[]
    menu.addItem("Preview", event -> {});
    menu.addItem("Edit", event -> {});
    menu.add(new Hr());

    // tag::snippet[]
    GridMenuItem<File> export = menu.addItem("Export");
    GridSubMenu<File> exportSubMenu = export.getSubMenu();
    exportSubMenu.addItem("Portable Document Format (.pdf)", event -> {});
    exportSubMenu.addItem("Rich Text Format (.rtf)", event -> {});
    exportSubMenu.addItem("Plain text (.txt)", event -> {});
    // end::snippet[]

    GridMenuItem<File> share = menu.addItem("Share");
    GridSubMenu<File> shareSubMenu = share.getSubMenu();
    shareSubMenu.addItem("Copy link", event -> {});
    shareSubMenu.addItem("Email", event -> {});
    menu.add(new Hr());

    menu.addItem("Delete", event -> {});

    add(grid);
  }

  private File[] getFiles() {
    return new File[] {
      new File("Annual Report.docx", 25165824),
      new File("Financials.xlsx", 44040192)
    };
  }

  private class File {

    private String name;
    private long size;

    File(String name, long size) {
      this.name = name;
      this.size = size;
    }

    public String getName() {
      return name;
    }

    public long getSize() {
      return size;
    }

    public String getDisplaySize() {
      return byteCountToDisplaySize(size);
    }
  }
  public static class Exporter extends DemoExporter<ContextMenuHierarchical> {} // hidden-source-line
}

File: frontend/demo/component/contextmenu/context-menu-hierarchical.ts
import 'Frontend/demo/init'; // hidden-source-line
import { html, LitElement } from 'lit';
import { customElement, state } from 'lit/decorators.js';
import '@vaadin/vaadin-context-menu/vaadin-context-menu';
import '@vaadin/vaadin-grid/vaadin-grid';
import type { GridElement, GridEventContext } from '@vaadin/vaadin-grid/vaadin-grid';
import { applyTheme } from 'Frontend/generated/theme';

interface FileItem {
  name: string;
  size: string;
}

@customElement('context-menu-hierarchical')
export class Example extends LitElement {
  protected createRenderRoot() {
    const root = super.createRenderRoot();
    // Apply custom theme (only supported if your app uses one)
    applyTheme(root);
    return root;
  }

  // tag::snippet[]
  @state()
  private items = [
    { text: 'Preview' },
    { text: 'Edit' },
    { component: 'hr' },
    {
      text: 'Export',
      children: [
        { text: 'Portable Document Format (.pdf)' },
        { text: 'Rich Text Format (.rtf)' },
        { text: 'Plain text (.txt)' },
      ],
    },
    { text: 'Share', children: [{ text: 'Copy link' }, { text: 'Email' }] },
    { component: 'hr' },
    { text: 'Delete' },
  ];
  // end::snippet[]

  @state()
  private gridItems: FileItem[] = [
    { name: 'Annual Report.docx', size: '24 MB' },
    { name: 'Financials.xlsx', size: '42 MB' },
  ];

  render() {
    return html`
      <!-- tag::snippethtml[] -->
      <vaadin-context-menu .items=${this.items}>
        <vaadin-grid
          all-rows-visible
          .items=${this.gridItems}
          @vaadin-contextmenu=${this.onContextMenu}
        >
          <vaadin-grid-column path="name"></vaadin-grid-column>
          <vaadin-grid-column path="size"></vaadin-grid-column>
        </vaadin-grid>
      </vaadin-context-menu>
      <!-- end::snippethtml[] -->
    `;
  }

  onContextMenu(e: MouseEvent) {
    // Prevent opening context menu on header row.
    const target = e.currentTarget as GridElement;
    if ((target.getEventContext(e) as GridEventContext<FileItem>).section !== 'body') {
      e.stopPropagation();
    }
  }
}
